### PR TITLE
v10 SQLite support + distributed locking abstractions

### PR DESCRIPTION
**Pull request by p-m-j**

### CODE REVIEW

**Comment by p-m-j** [comment]
Just integration tests at the moment.

File: src/Umbraco.Infrastructure/Persistence/NPocoDatabaseTypeExtensions.cs [hunk 1]
```diff
@@ -12,24 +14,16 @@ public static bool IsSqlServer(this DatabaseType databaseType)
...
-            return databaseType is NPoco.DatabaseTypes.SqlServer2008DatabaseType;
-        }
-
-        public static bool IsSqlServer2012OrLater(this DatabaseType databaseType)
-        {
-            return databaseType is NPoco.DatabaseTypes.SqlServer2012DatabaseType;
-        }
-
+        [Obsolete("Usage of this method indicates a code smell.")]
         public static bool IsSqlCe(this DatabaseType databaseType)
```
Comment (p-m-j) [inline src/Umbraco.Infrastructure/Persistence/NPocoDatabaseTypeExtensions.cs:18]: Not removed yet, discuss old migrations

File: src/Umbraco.Infrastructure/Persistence/Repositories/Implement/UserRepository.cs [hunk 2]
```diff
@@ -225,10 +225,10 @@ public bool ValidateLoginSession(int userId, Guid sessionId)
             // that query is going to run a *lot*, make it a template
             var t = SqlContext.Templates.Get("Umbraco.Core.UserRepository.ValidateLoginSession", s => s
                 .Select<UserLoginDto>()
-                .SelectTop(1)
                 .From<UserLoginDto>()
                 .Where<UserLoginDto>(x => x.SessionId == SqlTemplate.Arg<Guid>("sessionId"))
-                .ForUpdate());
+                .ForUpdate()
+                .SelectTop(1)); // Stick at end, SQL server syntax provider will insert at start of query after "select ", but sqlite will append limit to end.
```
Comment (p-m-j) [inline src/Umbraco.Infrastructure/Persistence/Repositories/Implement/UserRepository.cs:231]: Mention this gotya

File: src/Umbraco.Infrastructure/Persistence/ScalarMapper.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+using System;
+
+namespace Umbraco.Cms.Infrastructure.Persistence;
+
+public abstract class ScalarMapper<T> : IScalarMapper
```
Comment (p-m-j) [inline src/Umbraco.Infrastructure/Persistence/ScalarMapper.cs:5]: NPoco workaround sad times

File: src/Umbraco.Infrastructure/Persistence/SqlSyntax/SqlSyntaxProviderBase.cs [hunk 4]
```diff
@@ -245,14 +264,16 @@ public virtual IEnumerable<ColumnInfo> GetColumnsInSchema(IDatabase db)
...
         public abstract void ReadLock(IDatabase db, params int[] lockIds);
         public abstract void WriteLock(IDatabase db, params int[] lockIds);
+
+        public virtual string GetFieldNameForUpdate<TDto>(Expression<Func<TDto, object>> fieldSelector, string tableAlias = null) => this.GetFieldName(fieldSelector, tableAlias);
+
+        public virtual IDictionary<Type, IScalarMapper> ScalarMappers => null;
+
         public abstract void ReadLock(IDatabase db, TimeSpan timeout, int lockId);
 
         public abstract void WriteLock(IDatabase db, TimeSpan timeout, int lockId);
```
Comment (p-m-j) [inline src/Umbraco.Infrastructure/Persistence/SqlSyntax/SqlSyntaxProviderBase.cs:274]: Reminder
Lock timeouts will happen when attempting to begin transaction.
Umbraco code mostly assumes transaction begin immediately but time out.

File: src/Umbraco.Persistence.Sqlite/Services/SqliteBulkSqlInsertProvider.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        var inTrans = database.InTransaction;
+
+        if (!inTrans)
+        {
+            database.BeginTransaction();
+        }
+
+        foreach (var record in records)
+        {
+            database.Insert(record);
```
Comment (p-m-j) [inline src/Umbraco.Persistence.Sqlite/Services/SqliteBulkSqlInsertProvider.cs:44]: If we convert to query text we can instantiate params a single time and rebind

File: src/Umbraco.Persistence.Sqlite/Services/SqliteDatabaseCreator.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    /// A write transaction still requires exclusive access to database files so concurrent writes are not possible.
+    /// </para>
+    /// <para>
+    /// Read more <a href="https://www.sqlite.org/isolation.html">Isolation in SQLite</a> <br/>
+    /// Read more <a href="https://www.sqlite.org/wal.html">Write-Ahead Logging</a>
+    /// </para>
+    /// </remarks>
+    public void Create(string connectionString)
+    {
+        using var connection = new SqliteConnection(connectionString);
```
Comment (p-m-j) [inline src/Umbraco.Persistence.Sqlite/Services/SqliteDatabaseCreator.cs:36]: not settable on connection string for MS.Data.Sqlite

File: umbraco.sln [hunk 7]
```diff
@@ -132,82 +129,108 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Umbraco.Web.Common", "src\U
...
 		Debug|Any CPU = Debug|Any CPU
 		Release|Any CPU = Release|Any CPU
+		SkipTests|Any CPU = SkipTests|Any CPU
 	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
 		{DCDFE97C-5630-4F6F-855D-8AEEB96556A5}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
 		{DCDFE97C-5630-4F6F-855D-8AEEB96556A5}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{DCDFE97C-5630-4F6F-855D-8AEEB96556A5}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{DCDFE97C-5630-4F6F-855D-8AEEB96556A5}.Release|Any CPU.Build.0 = Release|Any CPU
-		{3819A550-DCEC-4153-91B4-8BA9F7F0B9B4}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
Comment (p-m-j) [inline umbraco.sln:147]: stops codeql kicking off about passwords in test projects etc

File: umbraco.sln [hunk 8]
```diff
@@ -132,82 +129,108 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Umbraco.Web.Common", "src\U
...
+		{79E4293D-C92C-4649-AEC8-F1EFD95BDEB1}.SkipTests|Any CPU.Build.0 = Debug|Any CPU
 		{2A5027D9-F71D-4957-929E-F7A56AA1B95A}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
 		{2A5027D9-F71D-4957-929E-F7A56AA1B95A}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{2A5027D9-F71D-4957-929E-F7A56AA1B95A}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{2A5027D9-F71D-4957-929E-F7A56AA1B95A}.Release|Any CPU.Build.0 = Release|Any CPU
-		{1B28FC3E-3D5B-4A46-8961-5483835548D7}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
-		{1B28FC3E-3D5B-4A46-8961-5483835548D7}.Debug|Any CPU.Build.0 = Debug|Any CPU
-		{1B28FC3E-3D5B-4A46-8961-5483835548D7}.Release|Any CPU.ActiveCfg = Release|Any CPU
-		{1B28FC3E-3D5B-4A46-8961-5483835548D7}.Release|Any CPU.Build.0 = Release|Any CPU
+		{2A5027D9-F71D-4957-929E-F7A56AA1B95A}.SkipTests|Any CPU.ActiveCfg = Debug|Any CPU
```
Comment (p-m-j) [inline umbraco.sln:220]: stops codeql kicking off about passwords in test projects etc

File: src/Umbraco.Persistence.Sqlite/Constants.cs [hunk 9]
```diff
@@ -8,5 +8,5 @@ public static class Constants
     /// <summary>
     /// SQLite provider name.
     /// </summary>
-    public const string ProviderName = "System.Data.SQLite";
+    public const string ProviderName = "Microsoft.Data.SQLite";
```
Comment (p-m-j) [inline src/Umbraco.Persistence.Sqlite/Constants.cs:11]: Discuss why

File: src/Umbraco.Infrastructure/Persistence/NPocoSqlExtensions.cs [hunk 10]
```diff
@@ -409,16 +409,34 @@ public static Sql<ISqlContext>.SqlJoinClause<ISqlContext> LeftJoin<TDto>(this Sq
...
-            var sqlJoin = sql.LeftJoin(join);
-            sql.Append(nestedSql);
-            return sqlJoin;
+            // HACK: SQLite
+            if (sql.SqlContext.DatabaseType.IsSqlite())
+            {
+                Sql<ISqlContext>.SqlJoinClause<ISqlContext> sqlJoin = sql.LeftJoin("(" + join);
+                sql.Append(nestedSql);
+                sql.Append($") {quotedAlias ?? tableName}");
+                return sqlJoin;
```
Comment (p-m-j) [inline src/Umbraco.Infrastructure/Persistence/NPocoSqlExtensions.cs:432]: move this mess to syntax provider

File: src/Umbraco.Persistence.SqlServer/Dtos/ColumnInSchemaDto.cs [hunk 11]
```diff
@@ -1,6 +1,6 @@
 using NPoco;
 
-namespace Umbraco.Cms.Infrastructure.Persistence.Dtos
+namespace Umbraco.Persistence.SqlServer.Dtos
 {
```
Comment (p-m-j) [inline src/Umbraco.Persistence.SqlServer/Dtos/ColumnInSchemaDto.cs:4]: Should be Umbraco.Cms.Persistence etc

File: src/Umbraco.Persistence.Sqlite/Constants.cs (hunk 9 above)
Comment (p-m-j) [reply src/Umbraco.Persistence.Sqlite/Constants.cs:11]: (it's because of apple silicon)

File: src/Umbraco.Persistence.Sqlite/Services/SqliteSyntaxProvider.cs [hunk 12]
```diff
@@ -0,0 +1,495 @@
...
+    private readonly IOptions<GlobalSettings> _globalSettings;
+    private readonly ILogger<SqliteSyntaxProvider> _log;
+    private readonly IDictionary<Type, IScalarMapper> _scalarMappers;
+
+    public SqliteSyntaxProvider(IOptions<GlobalSettings> globalSettings, ILogger<SqliteSyntaxProvider> log)
+    {
+        _globalSettings = globalSettings;
+        _log = log;
+
+        _scalarMappers = new Dictionary<Type, IScalarMapper>
```
Comment (p-m-j) [inline src/Umbraco.Persistence.Sqlite/Services/SqliteSyntaxProvider.cs:38]: should this move to DbProviderFactoryCreator

File: src/Umbraco.Infrastructure/DistributedLocking/InMemoryDistributedLockingMechanism.cs [hunk 13]
```diff
@@ -75,78 +77,94 @@ public IDistributedLock WriteLock(int lockId, TimeSpan? obtainLockTimeout)
...
+                distributedLock = InMemoryDistributedLock.Write(_parent, lockId);
+                _internalLocks.Add(distributedLock);
+                return true;
             }
 
-            _writeLocks.Add(lockId);
-            return true;
+            var existingLocks = _internalLocks.Where(x => x.LockId == lockId).ToList();
+            if (existingLocks.Count == 1 && existingLocks.First().ThreadId == Thread.CurrentThread.ManagedThreadId)
+            {
```
Comment (p-m-j) [inline src/Umbraco.Infrastructure/DistributedLocking/InMemoryDistributedLockingMechanism.cs:144]: We can do better if we make scope reader -> writer upgrades explicit rather than implied (but explicit upgrade is noop for sql server implementation).

Also tried with ReaderWriiterLockSlim but it's a bit meh as we don't know whether we need an upgradeable lock in advance so we kill concurrency (only single upgradeable lock at a time).

tl;dr. I'm not particularly happy with this setup but it's "OK" for now.

Might want to revisit this being the default impl when none is set in config.

**Comment by p-m-j** [comment]
> but I think we should extract the current name.

@bergmania doesn't seem worth it, strip |DataDirectory| guess if file extensions are relevant e.g. .mdf, .db etc.
Much easier to just hide the section if a connection string already configured.

File: src/Umbraco.Infrastructure/Persistence/UmbracoDatabaseFactory.cs [hunk 14]
```diff
@@ -248,9 +243,6 @@ private SqlContext Initialize()
                 {
                     cfg.WithInterceptor(interceptor);
                 }
-
-                cfg.WithInterceptor(new ProfiledDbConnectionInterceptor());
```
Comment (p-m-j) [inline src/Umbraco.Infrastructure/Persistence/UmbracoDatabaseFactory.cs]: Miniprofiler isn't really provider specific, but the order is important and this is a easy way to handle the ordering.
i.e. profiler should be before retry policy.

**Comment by p-m-j** [comment]
🎉 !!!
